homework/list.js

```diff
@@ -1,79 +1,146 @@
+        case 'click':
+          this.onNoteOpen(event);
+          break;
+      }
+    },
 
-  function updateList(list) {
-    _listNoteContent = list;
-  }
+    /**
+    * Initialize the listManager.
+    *
+    * @this {ListManager}
+    */
+    start() {
+      this._wrapper = document.querySelector('#note-list-wrapper');
+      this.fetchList()
+          .then(this.updateList.bind(this))
+          .then(this.drawList.bind(this))
+          .then(this.preloadFirstNote.bind(this))
+          .catch(function (error) {
+            console.log(error);
+          });
+      window.addEventListener('click', this);
+    },
 
-  function drawList() {
-    var list = _listNoteContent;
-    var ul = document.createElement('ul');
-    ul.id = 'note-title-list';
-    var buff = document.createDocumentFragment();
-    list.forEach(function(note, i) {
-      var li = document.createElement('li');
-      li.dataset.noteId = i;
-      li.classList.add('note-title');
-      li.textContent = note.title;
-      // Note: buff is captured, so we now have a
-      // little closure naturally.
-      buff.appendChild(li);
-    });
-    ul.appendChild(buff);
-    _wrapper.appendChild(ul);
-  }
+    /**
+    * Fetch the note list from server.
+    * In this demo case, we fetch data from localhost.
+    *
+    * @return {Promise} A start point of the work flow.
+    */
+    fetchList() {
+      return new Promise(function (resolve, reject) {
+        var xhr = new XMLHttpRequest();
+        xhr.open('GET', 'http://127.0.0.1:8080/demo-list-notes.json', true);
+        xhr.responseType = 'json';
+        xhr.onreadystatechange = (function (e) {
+          // Watch out: we have a mysterious unknown 'this'.
 
-  function fetchList(afterFetch) {
-    var xhr = new XMLHttpRequest();
-    xhr.open('GET', 'http://127.0.0.1:8000/demo-list-notes.json', true);
-    xhr.responseType = 'json';
-    xhr.onreadystatechange = function(e) {
-      // Watch out: we have a mysterious unknown 'this'.
-      if (this.readyState === 4 && this.status === 200) {
-        var listData = this.response;
-        // The flow ends here.
-        afterFetch(listData);
-      } else if (this.status !== 200 ){
-        // Ignore error in this case.
+          // Add bind xhr to specify or make code more clear. !?
+          if (this.readyState === 4 && this.status === 200) {
+            var listData = this.response;
+            // The flow ends here.
+            resolve(listData);
+          } else if (this.status !== 200) {
+            // Ignore error in this case.
+            reject({
+              status: this.status,
+              statusText: this.statusText
+            });
+          }
+        }).bind(xhr);
+        xhr.send();
+      });
+    },
+
+    /**
+    * When a note open, it will fire an 'note-open' event.
+    *
+    * @this {ListManager}
+    * @param {object} event The information such as event source, which is about event.
+    */
+    onNoteOpen(event) {
```

Comment:
It's a convention to prefix `on` to a method with the corresponding event. Therefore, this method should be renamed as `onClick`. It's a hidden issue about convention and expressiveness.